mirgecom/profiling.py

```diff
@@ -138,48 +92,113 @@ def __init__(self, queue, allocator=None) -> None:
+                    nbytes += arg.size * arg.dtype.itemsize
+                    nops += arg.size
+            res = SingleCallKernelProfile(time=0, flops=nops, bytes_accessed=nbytes,
+                                footprint_bytes=nbytes)
+            self.kernel_stats.setdefault(knl, {})[args_tuple] = res
+
+        if self.logmgr and f"{name}_time" not in self.logmgr.quantity_data:
+            self.logmgr.add_quantity(KernelProfile(self, name))
+
+        self.profile_events.append(ProfileEvent(evt, knl, args_tuple))
+
+        return evt
 
     def _finish_profile_events(self) -> None:
         # First, wait for completion of all events
         if self.profile_events:
             cl.wait_for_events([pevt.cl_event for pevt in self.profile_events])
 
-        global nonloopy_profile_events
-        if nonloopy_profile_events:
-            cl.wait_for_events([npe.cl_event for npe in nonloopy_profile_events])
-
         # Then, collect all events and store them
         for t in self.profile_events:
             program = t.program
             r = self._get_kernel_stats(program, t.args_tuple)
             time = t.cl_event.profile.end - t.cl_event.profile.start
 
-            new = ProfileResult(time, r.flops, r.bytes_accessed, r.footprint_bytes)
-
-            self.profile_results.setdefault(program, []).append(new)
+            new = SingleCallKernelProfile(time, r.flops, r.bytes_accessed,
+                                          r.footprint_bytes)
 
-        for t in nonloopy_profile_events:
-            program = t.knl
-            time = t.cl_event.profile.end - t.cl_event.profile.start
-            new = ProfileResult(time, t.nops, t.nbytes, t.nbytes)
             self.profile_results.setdefault(program, []).append(new)
 
         self.profile_events = []
-        nonloopy_profile_events = []
 
-    def tabulate_profiling_data(self) -> pytools.Table:
+    def get_and_reset_profiling_data_for_kernel(self, kernel_name: str,
+                                 wait_for_events=True) -> MultiCallKernelProfile:
+        """Return and reset profiling data for kernel `kernel_name`."""
+        if wait_for_events:
+            self._finish_profile_events()
+
+        num_calls = 0
+        time = 0
+        flops = 0
+        bytes_accessed = 0
+        fprint_bytes = 0
+
+        keys = [k for k in self.profile_results.keys() if k.name == kernel_name]
+
+        for key in keys:
+            value = self.profile_results[key]
+
+            num_calls += len(value)
+
+            time += sum([v.time for v in value])
+
+            if value[0].flops is not None:
+                flops += sum([v.flops for v in value])
+
+            if value[0].bytes_accessed is not None:
+                bytes_accessed += sum([v.bytes_accessed for v in value])
+
+            if value[0].footprint_bytes is not None:
+                fprint_bytes += sum([v.footprint_bytes for v in value])
+
+            del self.profile_results[key]
```

Comment:
Is this essential? Why do we need to consume/zero out `profile_results` (instead of letting it grow)? (If we'd like that functionality, could we make it a separate method?)

Reply:
I think the profile data will consume a lot of memory if we don't delete the old one. We would also need a mechanism to avoid getting stale data from a previous timestep (also to avoid iterating over longer and longer lists).

Reply:
It is *possible* to separate the "get" from the "reset" functionality while avoiding long lists: We could simply retain the computed `MultiCallKernelProfile` in `profile_results` and incorporate it/use it as a starting point in the next round of statistics. The bigger question is: do we want that? I could go both ways here.

Reply:
> It is _possible_ to separate the "get" from the "reset" functionality while avoiding long lists: We could simply retain the computed `MultiCallKernelProfile` in `profile_results` and incorporate it/use it as a starting point in the next round of statistics. The bigger question is: do we want that? I could go both ways here.

Hmm, not sure I understand. In the current implementation, a MultiCallKernelProfile object holds (mostly averaged) data for multiple executions of a single kernel during a single time step. Should the MultiCallKernelProfile span more than one time step? Or in other words, how would it be incorporated into the next round of statistics?

Reply:
> MultiCallKernelProfile object holds (mostly averaged) data for multiple executions of a single kernel during a single time step. Should the MultiCallKernelProfile span more than one time step

First off, this array context has no idea what a time step is, so the period over which statistics are kept is ultimately "from one reset to the next". I'm pushing for separating "reset" from "get statistics".

> Or in other words, how would it be incorporated into the next round of statistics?

If you have an average representing 10 data points, and you add one data point, you compute a 10:1 weighted average to compute the new average over 11 data points. In that sense, a `MultiCallKernelProfile` could simply represent multiple calls and could be incorporated into a new average by weighting.

Reply:
> First off, this array context has no idea what a time step is, so the period over which statistics are kept is ultimately "from one reset to the next". I'm pushing for separating "reset" from "get statistics".

Ok, fair. How would the "reset" get called from the logging infrastructure?

> > Or in other words, how would it be incorporated into the next round of statistics?
> 
> If you have an average representing 10 data points, and you add one data point, you compute a 10:1 weighted average to compute the new average over 11 data points. In that sense, a `MultiCallKernelProfile` could simply represent multiple calls and could be incorporated into a new average by weighting.

Ok.